pkg/statedb/reflector/k8s.go

```diff
@@ -0,0 +1,255 @@
+	ListerWatcher  cache.ListerWatcher  // The ListerWatcher to use to retrieve the objects
+	Transform      TransformFunc[Obj]   // Optional function to transform the objects given by the ListerWatcher
+	Table          statedb.RWTable[Obj] // The table to populate
+}
+
+// TransformFunc is an optional function to give to the Kubernetes source
+// to transform the object returned by the ListerWatcher to the desired
+// target object. If the function returns false the object is silently
+// skipped.
+type TransformFunc[Obj any] func(any) (obj Obj, ok bool)
+
+const (
+	DefaultBufferSize     = 64
+	DefaultBufferWaitTime = 50 * time.Millisecond
+)
+
+func (cfg KubernetesConfig[Obj]) getBufferSize() int {
+	if cfg.BufferSize == 0 {
+		return DefaultBufferSize
+	}
+	return cfg.BufferSize
+}
+
+func (cfg KubernetesConfig[Obj]) getWaitTime() time.Duration {
+	if cfg.BufferWaitTime == 0 {
+		return DefaultBufferWaitTime
+	}
+	return cfg.BufferWaitTime
+}
+
+type KubernetesParams[Obj any] struct {
+	cell.In
+
+	Config KubernetesConfig[Obj]
+	Jobs   job.Registry
+	Scope  cell.Scope
+	DB     *statedb.DB
+}
+
+// Kubernetes synchronizes statedb table with an external Kubernetes resource.
+// Returns a 'Source' for starting and stopping it. If the source can be started
+// unconditionally, consider using [KubernetesCell] instead.
+func Kubernetes[Obj any](p KubernetesParams[Obj]) Reflector[Obj] {
+	tr := &k8sSource[Obj]{
+		KubernetesConfig: p.Config,
+		db:               p.DB,
+	}
+	g := p.Jobs.NewGroup(p.Scope)
+	g.Add(job.OneShot("run", tr.run))
+	return g
+}
+
+// KubernetesCell constructs a cell that constructs the Kubernetes source and
+// adds it to the the application lifecycle.
+// For dependencies see [KubernetesParams].
+func KubernetesCell[Obj any]() cell.Cell {
+	return cell.Group(
+		cell.ProvidePrivate(Kubernetes[Obj]),
+		cell.Invoke(func(s Reflector[Obj], lc cell.Lifecycle) {
+			lc.Append(s)
+		}),
+	)
+}
+
+type k8sSource[Obj any] struct {
+	KubernetesConfig[Obj]
+
+	db *statedb.DB
+}
+
+func (s *k8sSource[Obj]) run(ctx context.Context, health cell.HealthReporter) error {
+	type entry struct {
+		deleted   bool
+		name      string
+		namespace string
+		obj       Obj
+	}
+	type buffer map[string]entry
+	bufferSize := s.getBufferSize()
+	waitTime := s.getWaitTime()
+	table := s.Table
+
+	transform := s.Transform
+	if transform == nil {
+		// No provided transform function, use the identity function instead.
+		transform = TransformFunc[Obj](func(obj any) (Obj, bool) { return obj.(Obj), true })
+	}
+
+	// Construct a stream of K8s objects, buffered into chunks every 100 milliseconds.
+	// This reduces the number of write transactions required and thus the number of times
+	// readers get woken up, which results in much better overall throughput.
+	src := stream.Buffer(
+		k8sEventObservable(s.ListerWatcher),
+		bufferSize,
+		waitTime,
+
+		// Buffer the events into a map, coalescing them by key.
```

Comment:
1. Are we guaranteed to get the events in the right order? As in, we cannot get a sequence like `Add, Update2, Update1` and end up with the wrong state?

Reply:
Yep, `cache.Reflector` will do the store operations in the correct order and thus we get the events in the right order.